File: 2020/day11/main.cpp
#include <algorithm>
#include <cassert>
#include <fstream>
#include <iostream>
#include <vector>

using std::count;
using std::cout;
using std::endl;
using std::ifstream;
using std::string;
using std::vector;

using Grid = vector<vector<char>>;

int count(const Grid& grid) {
    int ret = 0;
    for (const auto& r: grid) {
        ret += (int) count(r.begin(), r.end(), '#');
    }
    return ret;
}

Grid next_adjacent(const Grid& grid) {
    int rows = (int) grid.size();
    int cols = (int) grid[0].size();
    Grid ret;
    for (int row = 0; row < rows; row++) {
        vector<char> r;
        for (int col = 0; col < cols; col++) {
            if (grid[row][col] == '.') {
                r.push_back('.');
                continue;
            }
            int count = 0;
            for (int nr = row - 1; nr <= row + 1; nr++) {
                for (int nc = col - 1; nc <= col + 1; nc++) {
                    if ((nr != row || nc != col) && (0 <= nr && nr < rows && 0 <= nc && nc < cols)) {
                        count += grid[nr][nc] == '#';
                    }
                }
            }
            if (count == 0) {
                r.push_back('#');
            } else if (count >= 4) {
                r.push_back('L');
            } else {
                r.push_back(grid[row][col]);
            }
        }
        ret.push_back(r);
    }
    return ret;
}

Grid next_line_of_sight(const Grid& grid) {
    int rows = (int) grid.size();
    int cols = (int) grid[0].size();
    Grid ret;
    for (int row = 0; row < rows; row++) {
        vector<char> r;
        for (int col = 0; col < cols; col++) {
            if (grid[row][col] == '.') {
                r.push_back('.');
                continue;
            }
            int count = 0;
            for (int dr = -1; dr <= +1; dr++) {
                for (int dc = -1; dc <= +1; dc++) {
                    if (dr == 0 && dc == 0) {
                        continue;
                    }
                    int i = 1;
                    while (true) {
                        int nr = row + dr * i;
                        int nc = col + dc * i;
                        if (!(0 <= nr && nr < rows && 0 <= nc && nc < cols)) {
                            break;
                        }
                        char c = grid[nr][nc];
                        if (c == '#') {
                            count++;
                            break;
                        } else if (c == 'L') {
                            break;
                        }
                        i++;
                    }
                }
            }
            if (count == 0) {
                r.push_back('#');
            } else if (count >= 5) {
                r.push_back('L');
            } else {
                r.push_back(grid[row][col]);
            }
        }
        ret.push_back(r);
    }
    return ret;
}

int puzzle1(Grid grid) {
    while (true) {
        Grid new_grid = next_adjacent(grid);
        if (new_grid == grid) {
            return count(grid);
        }
        grid = new_grid;
    }
}

int puzzle2(Grid grid) {
    while (true) {
        Grid new_grid = next_line_of_sight(grid);
        if (new_grid == grid) {
            return count(grid);
        }
        grid = new_grid;
    }
}

int main(void) {
    ifstream f("input");
    assert(f.is_open());
    Grid grid;
    string line;
    while (getline(f, line)) {
        grid.emplace_back(line.begin(), line.end());
    }

    cout << puzzle1(grid) << endl;
    cout << puzzle2(grid) << endl;
}

File: 2020/day11/main.py
from typing import Callable, List

Grid = List[List[str]]


def count_adjacent(grid: Grid, row: int, col: int) -> int:
    rows = len(grid)
    cols = len(grid[0])
    return sum(
        grid[nr][nc] == '#'
        for nr in (row - 1, row, row + 1)
        for nc in (col - 1, col, col + 1)
        if rows > nr >= 0 <= nc < cols and (nr, nc) != (row, col)
    )


def count_lineofsight(grid: Grid, row: int, col: int) -> int:
    rows = len(grid)
    cols = len(grid[0])
    count = 0
    for dr in (-1, 0, +1):
        for dc in (-1, 0, +1):
            if dr == dc == 0:
                continue
            i = 1
            while True:
                nr = row + dr * i
                nc = col + dc * i
                if not rows > nr >= 0 <= nc < cols:
                    break
                if grid[nr][nc] == '#':
                    count += 1
                    break
                elif grid[nr][nc] == 'L':
                    break
                i += 1
    return count


def step(
    grid: Grid,
    counter: Callable[[Grid, int, int], int],
    max_neighbors: int
) -> Grid:
    rows = len(grid)
    cols = len(grid[0])

    ret = []
    for row in range(rows):
        r = []
        for col in range(cols):
            if grid[row][col] == '.':
                r.append('.')
                continue
            count = counter(grid, row, col)
            if count == 0:
                r.append('#')
            elif count > max_neighbors:
                r.append('L')
            else:
                r.append(grid[row][col])
        ret.append(r)
    return ret


def count_final(
    grid: Grid,
    counter: Callable[[Grid, int, int], int],
    max_neighbors: int,
) -> int:
    while True:
        new_grid = step(grid, counter, max_neighbors)
        if new_grid == grid:
            return sum(line.count('#') for line in grid)
        grid = new_grid


def main() -> None:
    with open('input') as f:
        grid = [list(line.strip()) for line in f]

    print(count_final(grid, count_adjacent, 3))  # puzzle 1
    print(count_final(grid, count_lineofsight, 4))  # puzzle 2


if __name__ == '__main__':
    main()
